util: implement util.getSystemErrorName()

joyeecheung opened the pull request:
Reimplement uv.errname() as internal/util.getErrorName() to
avoid the memory leaks caused by unknown error codes
and avoid calling into C++ for the error names. Also
expose it as a public API for external use.

The next step would be to deprecate `uv.errname()` (since the binding seems to be used in the user land, e.g. [execa](https://github.com/sindresorhus/execa/blob/80a059d/lib/errname.js)), remove the C++ code and monkey patch the binding to `internal/util.getErrorName()` during bootstrapping, although I am not sure when during the bootstrapping process is the best time to do that.

Refs: http://docs.libuv.org/en/v1.x/errors.html#c.uv_err_name

> const char* uv_err_name(int err)
>   Returns the error name for the given error code. Leaks a few bytes of memory when you call it with an unknown error code
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
util, child_process, test

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12562/

jasnell commented on lib/internal/util.js:
```diff
@@ -213,6 +214,14 @@ function getConstructorOf(obj) {
+    return errmap.get(code)[0];
+  } else {
+    return `Unknown system error ${code}`;
+  }
```
Rather than the separate `has()` and `get()` calls, perhaps just

```js
function getErrorName(code) {
  const entry = errmap.get(code);
  return entry ? entry[0] : `Unknown system error ${code}`;
}
```

jasnell commented on lib/internal/util.js:
```diff
@@ -213,6 +214,14 @@ function getConstructorOf(obj) {
 }
 
+function getErrorName(code) {
+  if (errmap.has(code)) {
```
should `code` be type checked at all?

joyeecheung replied:
@jasnell Yeah, `errname()` does `CHECK_LT(err, 0)`, we should probably check for the type and the ranges. Thanks for catching that.

vsemozhetbyt commented on doc/api/util.md:
```diff
@@ -1362,6 +1381,7 @@ Deprecated predecessor of `console.log`.
 [Customizing `util.inspect` colors]: #util_customizing_util_inspect_colors
 [Internationalization]: intl.html
 [WHATWG Encoding Standard]: https://encoding.spec.whatwg.org/
+[Common System Errors]: errors.html#common_system_errors
```
`#common_system_errors` -> `#errors_common_system_errors` )

vsemozhetbyt commented on doc/api/util.md:
````diff
@@ -257,6 +257,25 @@ intended as a debugging tool. Some input values can have a significant
+```js
+fs.access('file/that/does/not/exist', (err) => {
+  const name = util.getErrorName(err.errno);
+  console.log(name);  // ENOENT
````
Maybe `console.error(name)`?

joyeecheung commented:
@addaleax @jasnell @vsemozhetbyt Thanks for the reviews. A few updates:

- `getErrorName()` ->  `getSystemErrorName()`
- Simplified the error map access logic
- Move the typechecking of `_errnoException()` into `getSystemErrorName()`, renamed the argument to `err` to be consistent with `_errnoException()`
- Fixed the link in the documentation

PTAL

New CI: https://ci.nodejs.org/job/node-test-pull-request/12564/

jasnell commented on lib/internal/util.js:
```diff
@@ -213,6 +214,19 @@ function getConstructorOf(obj) {
+  if (typeof err !== 'number') {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'err', 'number', err);
+  }
+  if (err >= 0 || !Number.isSafeInteger(err)) {
```
I believe that `if (err >> 0 |== err)` should work also

jasnell commented:
Still LGTM :-)

BridgeAR commented on doc/api/util.md:
````diff
@@ -257,6 +257,25 @@ intended as a debugging tool. Some input values can have a significant
+```js
+fs.access('file/that/does/not/exist', (err) => {
+  const name = util.getSystemErrorName(err.errno);
+  console.error(name);  // ENOENT
````
Suggestion: It might be good to also show the `err.errno` here to outline the difference.

joyeecheung replied:
@Bridge The numeric code is platform-dependent so I am not sure if it's OK to just show the value of a particular platform or should I need to show all the possible values...

BridgeAR replied:
I guess you meant me ;-)

And true, I forgot about that. Hm, either a "e.g. on platform xyz" or just land it as is. What ever you think is best.

joyeecheung commented:
Landed in 4af1bba6d0f1da1b4d444ef3a936e2ef2c639a1e, thanks!

joyeecheung commented:
Note to reporters: this needs to land with c56972779b8c239a48df46b7caadf1ef1b9eacd2 which belongs to a semver-major PR. I will see if I can manually backport this.

MylesBorins commented:
setting backport label
